### PR TITLE
[RFC] Lift freqs_cis as an input of models

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #1857
* __->__ #1797
* #1939

freqs_cis is sensitive to the sequence order. CP load balancing will shuffle the samples, so each batch will have different orders.  As a result, we will have to lift these order sensitive buffer to the inputs and broadcast them along the batch dimension so that PP will correctly shard freqs_cis without messing up the correctness.

Next step: once the design is finalized and people are okay with the change, we will extend this change to all models.

### CODE REVIEW

File: torchtitan/protocols/model.py [hunk 1]
```diff
@@ -70,3 +70,11 @@ def get_attention_masks(
         raise NotImplementedError(
             "This model does not support attention masking/Flex Attention."
         )
+
+    def get_order_sensitive_buffers(
+        self,
+        batch_size: int,
+        seq_len: int,
+    ) -> tuple[tuple[torch.Tensor, ...], tuple[int, ...]]:
```
Comment (wwwjn) [inline torchtitan/protocols/model.py]: Can we add some notes here what does this 2 return values mean? Seems like the first return value is the buffer itself

File: torchtitan/protocols/model.py [hunk 2]
```diff
@@ -70,3 +70,11 @@ def get_attention_masks(
         raise NotImplementedError(
             "This model does not support attention masking/Flex Attention."
         )
+
+    def get_order_sensitive_buffers(
```
Comment (tianyu-l) [inline torchtitan/protocols/model.py:74]: Naming is a bit vague. I think here we are only targeting "sequence dim" order-sensitive buffers, not the batch dim.

File: torchtitan/models/llama3/model/model.py [hunk 3]
```diff
@@ -437,9 +436,18 @@ def get_attention_masks(
             mask_mod, B, None, input_batch.shape[1], input_batch.shape[1]
         )
 
+    def get_order_sensitive_buffers(
+        self,
+        batch_size: int,
+        seq_len: int,
+    ) -> tuple[tuple[torch.Tensor, ...], tuple[int, ...]]:
+        freqs_cis = self.freqs_cis[:seq_len].repeat(batch_size, 1, 1)
```
Comment (tianyu-l) [inline torchtitan/models/llama3/model/model.py:481]: I wonder what's the benefit of keeping `self.freqs_cis`.

If seq len changes from iteration to iteration (e.g. in forge), it might be good to keep a central `self.freqs_cis` instead of computing it each iteration. The other benefit is that we may not want torchtitan model definition to deviate from "original" / "conventional" model definitions too much.

On the other hand, the dependency sounds indirect and error-prone:
- we create `self.freqs_cis` in model code
- then copy it to `freqs_cis`, which technically is outside the model
- we then send `freqs_cis` into model

Would like to hear your thoughts.

**Review by tianyu-l — COMMENTED** [review]
Since this is torchtitan-specific way of defining models, would be good if we can add a section in https://github.com/pytorch/torchtitan/blob/main/docs/composability.md

File: torchtitan/models/llama3/model/model.py (hunk 3 above)
Comment (fegin) [reply torchtitan/models/llama3/model/model.py:481]: Re-computation is the main reason why I decided to keep `self.freqs_cis`. I agree it's a bit awkward. 

One alternative is to sill keep `self.freqs_cis`  but set it as an optional field (`self.freqs_cis: torch.Tensor | None`) for bookkeeping only. And we only initialize it in this function. So the creation logic flow  (precompute and slicing) is mainly in this function. The model code still provides precompute function. So this way we do not change the code structure too much while keeping the logic together. Not a perfect solution though.

Comment (tianyu-l) [reply torchtitan/models/llama3/model/model.py:481]: I think it's fine to keep the current way for now, as it sounds more lightweight change, and as I mentioned downstream application (e.g. forge, and simple generation) may change `seq_len` from iteration to iteration, where we can avoid recomputation this way.

Comment (tianyu-l) [inline torchtitan/models/llama3/model/model.py:481]: In the (maybe unlikely) case when both seq len and batch size are large, this code is creating a big buffer (which may not peak the memory anyway?).
Might be OK as in reality for ultra-long seq len we should probably set batch size = 1.

**Review by tianyu-l — COMMENTED** [review]
SGTM, may need more benchmarking to verify the value of load balancing, which is the motivation of this change.
